File: ruby/lib/testivai/shard.rb
# frozen_string_literal: true

require "json"
require "time"
require "pathname"

module Testivai
  # Shard participation for parallel and sharded runs.
  #
  # Playwright can tell its reporter `--shard=i/N`; RSpec cannot. So the
  # mechanism is the environment, not a framework API, and every TestivAI
  # adapter honours the same two variables:
  #
  #   TESTIVAI_CAPTURE_ONLY=1   capture, do not compare
  #   TESTIVAI_SHARD=3/8        this process is shard 3 of 8
  #
  # That is what keeps a Ruby suite a first-class citizen of the same sharded
  # CI flow as Playwright rather than a second-class one.
  module Shard
    PATTERN = %r{\A\s*(\d+)\s*(?:/|of)\s*(\d+)\s*\z}i
    MANIFEST = "testivai-shard.json"

    module_function

    # Parse TESTIVAI_SHARD; nil when unset or malformed.
    def from_env(env = ENV)
      parse(env["TESTIVAI_SHARD"])
    end

    def parse(raw)
      m = PATTERN.match(raw.to_s)
      return nil unless m

      current = m[1].to_i
      total = m[2].to_i
      return nil if total < 1 || current < 1 || current > total

      { current: current, total: total }
    end

    def capture_only?(env = ENV)
      raw = env["TESTIVAI_CAPTURE_ONLY"].to_s
      return !%w[0 false].include?(raw.downcase) unless raw.empty?

      s = from_env(env)
      !s.nil? && s[:total] > 1
    end

    # Write (or refresh) the manifest merge-captures reads.
    #
    # RSpec gives the gem no end-of-run hook, so this is refreshed on every
    # capture and `complete` stays false — it records that the shard
    # participated, not that it finished.
    def write_manifest(temp_dir, shard, complete: false)
      dir = Pathname.new(temp_dir)
      dir.mkpath
      captures = dir.children.select(&:directory?).map { |p| p.basename.to_s }.sort
      dir.join(MANIFEST).write(
        JSON.pretty_generate(
          "shard" => { "current" => shard[:current], "total" => shard[:total] },
          "captures" => captures,
          "complete" => complete,
          "timestamp" => Time.now.utc.iso8601
        )
      )
    rescue StandardError # rubocop:disable Lint/SuppressedException
      # Never fail a capture over a manifest.
    end
  end
end

File: java/src/main/java/ai/testiv/testivai/Shard.java
package ai.testiv.testivai;

import com.google.gson.Gson;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Shard participation for parallel and sharded runs.
 *
 * <p>Playwright can tell its reporter {@code --shard=i/N}; JUnit cannot. So the
 * mechanism is the environment, not a framework API, and every TestivAI adapter
 * honours the same two variables:
 *
 * <pre>
 *   TESTIVAI_CAPTURE_ONLY=1   capture, do not compare or write a report
 *   TESTIVAI_SHARD=3/8        this process is shard 3 of 8
 * </pre>
 *
 * <p>That is what keeps a Java suite a first-class citizen of the same sharded
 * CI flow as Playwright rather than a second-class one.
 */
final class Shard {

  private static final Pattern PATTERN = Pattern.compile("\\s*(\\d+)\\s*(?:/|of)\\s*(\\d+)\\s*", Pattern.CASE_INSENSITIVE);
  private static final Gson GSON = new Gson();

  final int current;
  final int total;

  private Shard(int current, int total) {
    this.current = current;
    this.total = total;
  }

  /** Parse {@code TESTIVAI_SHARD}; null when unset or malformed. */
  static Shard fromEnv() {
    return parse(System.getenv("TESTIVAI_SHARD"));
  }

  static Shard parse(String raw) {
    if (raw == null || raw.isBlank()) return null;
    Matcher m = PATTERN.matcher(raw);
    if (!m.matches()) return null;
    int current = Integer.parseInt(m.group(1));
    int total = Integer.parseInt(m.group(2));
    if (total < 1 || current < 1 || current > total) return null;
    return new Shard(current, total);
  }

  /** Explicit env wins; otherwise being one shard of many implies capture-only. */
  static boolean captureOnly() {
    String raw = System.getenv("TESTIVAI_CAPTURE_ONLY");
    if (raw != null && !raw.isBlank()) {
      return !raw.equals("0") && !raw.equalsIgnoreCase("false");
    }
    Shard s = fromEnv();
    return s != null && s.total > 1;
  }

  /** Write the manifest {@code merge-captures} reads to prove every shard reported. */
  static void writeManifest(Path tempDir, Shard shard, boolean complete) {
    try {
      Files.createDirectories(tempDir);
      List<String> captures = new ArrayList<>();
      try (Stream<Path> entries = Files.list(tempDir)) {
        entries.filter(Files::isDirectory).forEach(p -> captures.add(p.getFileName().toString()));
      }
      Collections.sort(captures);

      String json =
          GSON.toJson(
              Map.of(
                  "shard", Map.of("current", shard.current, "total", shard.total),
                  "captures", captures,
                  "complete", complete,
                  "timestamp", Instant.now().toString()));
      Files.writeString(tempDir.resolve("testivai-shard.json"), json, StandardCharsets.UTF_8);
    } catch (IOException | RuntimeException ignored) {
      // A manifest we cannot write must never fail a test run.
    }
  }
}
